### PR TITLE
Advance

### PR DESCRIPTION
**Pull request by ar-maged**
**Features:**
- Basic signup & login system is intergrated using GraphQL, Mongoose & JWT
- Podcast search by genre _(iTunes)_ & category _(Pocket Casts)_
- Episodes pagination

**Usability & readability:**
- All constants, variables, arguments & returns are typed
- Some variables & functions are renamed
- GraphQL resolvers are factored out into separate functions
- All code is beautified & conforming to style guide
- `README.md` contains GraphQL querying use cases

**Fixes:**
- All GraphQL schema fields are nullable

**Limitations:**
- Podcast search by category returns a reduced podcast type containing only name & large artwork fields

### CODE REVIEW

**Comment by ar-maged** [comment]
@ahmedlhanafy Apologies for the lengthy pull request :pray:

File: api/itunes/connectors.ts [hunk 1]
```diff
@@ -1,39 +1,73 @@
...
+  return normalizeCategoricalPodcasts(trendingPodcasts);
+};
+
+export const getPopularPodcasts = async (): Promise<Array<PodcastAPI>> => {
+  const url: string = new PocketCastsUrlBuilder().popular().toString();
+  const popularPodcasts: any = await fetchCategoricalPodcasts({ url });
+  return normalizeCategoricalPodcasts(popularPodcasts);
+};
+
+export const fetchEpisodes = async ({ feedUrl, first, offset }:
```
Comment (ahmedlhanafy) [inline api/itunes/connectors.ts]: `offset` should defaults to zero, consider this case 
```graphql
{
  podcasts(name: "Modern", limit: 1) {
    episodes(first: 1) {
      title
    }
  }
}
```

Comment (ahmedlhanafy) [inline api/itunes/connectors.ts]: This will give wrong response

File: api/graphql/resolvers.ts [hunk 2]
```diff
@@ -1,118 +1,172 @@
...
+const resolvePodcasts = async ({ id, name, genre, category, limit }:
+  { id: string, name: string, genre: string, category: string, limit: number }):
+  Promise<Array<PodcastAPI>> => {
+  let results: Array<PodcastAPI>;
+  if (id) {
+    results = await findOnePodcast({ id });
+  } else {
+    if (category) {
+      switch (category) {
+        case 'FEATURED': results = await getFeaturedPodcasts(); break;
```
Comment (ahmedlhanafy) [inline api/graphql/resolvers.ts]: These functions should take `limit` as a param and limit the results accordingly

File: api/graphql/resolvers.ts [hunk 3]
```diff
@@ -1,118 +1,172 @@
...
+        case 'TRENDING': results = await getTrendingPodcasts(); break;
```
Comment (ahmedlhanafy) [inline api/graphql/resolvers.ts]: ^

File: api/graphql/resolvers.ts [hunk 4]
```diff
@@ -1,118 +1,172 @@
...
+        case 'POPULAR': results = await getPopularPodcasts(); break;
```
Comment (ahmedlhanafy) [inline api/graphql/resolvers.ts]: ^

File: api/itunes/connectors.ts [hunk 5]
```diff
@@ -1,39 +1,73 @@
...
 };
 
-export const searchEpisodes = async ({ feedUrl, limit }:
-  { feedUrl: string, limit?: number }): Promise<Array<Episode>> => {
-  const podcastData = await fetch(feedUrl);
-  const podcastJsonData = await podcastData.text();
-  const podcastParsedData = await parsePodcastPromise(podcastJsonData);
-  return podcastParsedData.episodes
-    .slice(0, limit || podcastParsedData.episodes.length - 1);
+const fetchCategoricalPodcasts = async ({ url }: { url: string }):
```
Comment (ahmedlhanafy) [inline api/itunes/connectors.ts]: This doesn't work but this is a known bug or limitation, right?
```graphql
{
  podcasts(category: POPULAR) {
    feedUrl
  }
}
```

Comment (ar-maged) [inline api/itunes/connectors.ts]: Yes, it is a limitation. We are able to request `FEATURED`, `TRENDING` & `POPULAR` podcasts from the PocketCasts API, but feed urls are not part of that response.

File: api/itunes/connectors.ts (hunk 1 above)
Comment (ar-maged) [inline api/itunes/connectors.ts]: Astute observation :zap:

**Pull request merged**